dali/util/numpy.cc

```diff
@@ -0,0 +1,218 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/numpy.h"
+#include <string>
+#include <vector>
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+namespace numpy {
+
+const TypeInfo &TypeFromNumpyStr(const std::string &format) {
+  if (format == "u1") return TypeTable::GetTypeInfo<uint8_t>();
+  if (format == "u2") return TypeTable::GetTypeInfo<uint16_t>();
+  if (format == "u4") return TypeTable::GetTypeInfo<uint32_t>();
+  if (format == "u8") return TypeTable::GetTypeInfo<uint64_t>();
+  if (format == "i1") return TypeTable::GetTypeInfo<int8_t>();
+  if (format == "i2") return TypeTable::GetTypeInfo<int16_t>();
+  if (format == "i4") return TypeTable::GetTypeInfo<int32_t>();
+  if (format == "i8") return TypeTable::GetTypeInfo<int64_t>();
+  if (format == "f2") return TypeTable::GetTypeInfo<float16>();
+  if (format == "f4") return TypeTable::GetTypeInfo<float>();
+  if (format == "f8") return TypeTable::GetTypeInfo<double>();
+  DALI_FAIL("Unknown Numpy type string");
+}
+
+inline void SkipSpaces(const char*& ptr) {
+  while (::isspace(*ptr))
+    ptr++;
+}
+
+template <size_t N>
+void Skip(const char*& ptr, const char (&what)[N]) {
+  DALI_ENFORCE(!strncmp(ptr, what, N - 1),
+               make_string("Expected \"", what, "\" but got \"", ptr, "\""));
+  ptr += N - 1;
+}
+
+template <size_t N>
+bool TrySkip(const char*& ptr, const char (&what)[N]) {
+  if (!strncmp(ptr, what, N-1)) {
+    ptr += N - 1;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template <size_t N>
+void SkipFieldName(const char*& ptr, const char (&name)[N]) {
+  SkipSpaces(ptr);
+  Skip(ptr, "'");
+  Skip(ptr, name);
+  Skip(ptr, "'");
+  SkipSpaces(ptr);
+  Skip(ptr, ":");
+  SkipSpaces(ptr);
+}
+
+template <typename T = int64_t>
+T ParseInteger(const char*& ptr) {
+  char *out_ptr = const_cast<char*>(ptr);  // strtol takes a non-const pointer
+  T value = static_cast<T>(strtol(ptr, &out_ptr, 10));
+  DALI_ENFORCE(out_ptr != ptr, "Parse error: expected a number.");
+  ptr = out_ptr;
+  return value;
+}
+
+std::string ParseStringValue(const char*& input, char delim_start = '\'', char delim_end = '\'') {
+  DALI_ENFORCE(*input++ == delim_start, make_string("Expected \'", delim_start, "\'"));
+  std::string out;
+  for (; *input != '\0'; input++) {
+    if (*input == '\\') {
+      switch (*++input) {
+        case '\\':
+          out += '\\';
+          break;
+        case '\'':
+          out += '\'';
+          break;
+        case '\t':
+          out += '\t';
+          break;
+        case '\n':
+          out += '\n';
+          break;
+        case '\"':
+          out += '\"';
+          break;
+        default:
```

Comment:
Nitpick: I guess that if it's a proper string, then I guess it should also support `\0oct` and `\xhex`.

Reply:
At least not in this PR, but maybe we should take care of it if we decide to do some serious numpy reader refactoring